compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Alpha.hs

```diff
@@ -369,6 +369,7 @@ alphaUpdate env = \case
             && alphaExpr' env e1a e2a
             && alphaExpr' env e1b e2b
         _ -> False
+    USoftExercise{} ->  undefined -- TODO https://github.com/digital-asset/daml/issues/16151
```

Comment:
this should be the same as `UExercise`, modulo constructor name 
```suggestion
    USoftExercise t1 c1 e1a e1b -> \case
        USoftExercise t2 c2 e2a e2b -> alphaTypeCon t1 t2
            && c1 == c2
            && alphaExpr' env e1a e2a
            && alphaExpr' env e1b e2b
        _ -> False
```

Reply:
It was a deliberate choice to leave these `undefined`, for now, to make it clear that so far we have no tests/examples which hit this code path. However, since the implementation is in some sense _purely_ structural (and identical to the non-soft case), then perhaps we can go ahead and add the implementation now.... I will do this
.

Reply:
done